### PR TITLE
test-bot: revise Step output transcoding and XML character filtering

### LINKED ISSUE #16: Jenkins runs fail on invalid XML: "Invalid byte 2 of 4-byte UTF-8 sequence"
Some Jenkins build runs will fail completely without output like this. No test results or artifacts will be available.

```
Failed to read test report file /Users/brew/Jenkins/prs/version/yosemite/brew-test-bot.xml
org.dom4j.DocumentException: Invalid byte 2 of 4-byte UTF-8 sequence. Nested exception: Invalid byte 2 of 4-byte UTF-8 sequence.
    at org.dom4j.io.SAXReader.read(SAXReader.java:484)
    at org.dom4j.io.SAXReader.read(SAXReader.java:264)
    at hudson.tasks.junit.SuiteResult.parse(SuiteResult.java:123)
    at hudson.tasks.junit.TestResult.parse(TestResult.java:286)
    at hudson.tasks.junit.TestResult.parsePossiblyEmpty(TestResult.java:232)
    at hudson.tasks.junit.TestResult.parse(TestResult.java:167)
```

The test output is all included in an XML file, which includes the output of some of the steps. That output, from commands we call, may not be valid XML CDATA, or even valid UTF-8, since we're calling external commands.

The `test-bot` command [attempts to filter out invalid characters](https://github.com/Homebrew/brew/blob/ac8088304603a95e2ae4d187016a50c36c622605/Library/Homebrew/cmd/test-bot.rb#L945-L946) but it's apparently not being thorough enough. We need more robust filtering or escaping.

I think the approach in https://github.com/Homebrew/legacy-homebrew/pull/50092 was pretty solid, but needed different syntax, like so.

```
output = step.output.gsub(/[^\x09\x0A\x0D\x20-\uD7FF\uE000-\uFFFD\u{10000}-\u{10FFFF}]/, "")
```

And it might be better to replace filtered-out characters with a Unicode "replacement character" placeholder instead of just eliding. Then there'd still be some indication that there was other output there.

###  References 

* Blocking https://github.com/Homebrew/legacy-homebrew/pull/49971
* https://github.com/Homebrew/legacy-homebrew/pull/50092 was an attempt to fix, but didn't work out
 * a6b918d6c5cd8a05494002d288f15d9b7b30f9df – original
 * 60055ba8901dfe2844c323101d2e425ce800cf2e – fixup
 * 161a1e286ec238cc9449a1d3e40362ce49e207bc – reverted to old code
* Caused breakage in https://github.com/Linuxbrew/linuxbrew/pull/1037#event-612849650

### PR DESCRIPTION
**Pull request by apjanke**
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031). *Answer*: nope; couldn't figure out how to put test-bot in a test harness or isolate this code.
- [X] Have you successfully ran `brew tests` with your changes locally?

##  Description

This replaces #24 and has identical changes. Something went wonky when I tried to push amendments to that PR, so I'm opening a new one.

Fixes #16.

This takes care of the occasional "Invalid XML" or "Invalid UTF-8" errors we see on Jenkins, where it cannot read

This change alters `test-bot` to:
  * Buffer the steps' output as binary, and only convert it to UTF-8 once collected, to avoid transcoding problems with multibyte characters.
  * Use a more robust regexp to filter out characters that are not allowed in XML. Instead of excluding some known-bad characters, it lists the known-valid character ranges and excludes everything else.
  * Replace filtered-out characters with the `xFFFD` Unicode replacement character instead of just deleting them, so it's obvious when there was junk in a test step's output.
  * Replace the single-byte `read(1)` with `readpartial` and a larger buffer, for performance.
  * Handle an edge case in the truncation code's size calculation.
  * Fix `--keep-logs` on Ruby 1.8.7, which lacks `File.write`

### CODE REVIEW

File: Library/Homebrew/cmd/test-bot.rb [hunk 1]
```diff
@@ -944,11 +953,22 @@ def test_bot
...
+            end
 
             if output.bytesize > BYTES_IN_1_MEGABYTE
-              output = "truncated output to 1MB:\n" \
-                + output.slice(-BYTES_IN_1_MEGABYTE, BYTES_IN_1_MEGABYTE)
+              slice_start = [0, output.length - BYTES_IN_1_MEGABYTE].max
+              if slice_start > 0
+                output = "truncated output to ~1MB:\n" \
+                  + output.slice(slice_start, output.length - slice_start)
+              end
```
Comment (UniqMartin) [inline Library/Homebrew/cmd/test-bot.rb]: I'm afraid my previous comment about `length != bytesize` (unless we have pure ASCII) was lost. Let's assume we have a string `x` with `x.length == 1_500_000` and `x.bytesize == 1_800_000` (assuming 20% of code units need two bytes in UTF-8 encoding). If I'm reading the code correctly, `slice` will operate on characters, start at character 451,424, and copy 1,048,576 characters. Still assuming the overhead from multi-byte code units, this will result in 1,258,291 bytes of output (significantly more than the desired 1 MB).

Admittedly, all of that is somewhat constructed and 20% is a big percentage given the output we typically capture. Still could be relevant depending on how sharp this 1 MB limit is.

An easy workaround could be: Truncate first while the encoding is still `BINARY` and convert to UTF-8 (with filtering/fixing the output) afterwards.

Comment (apjanke) [inline Library/Homebrew/cmd/test-bot.rb]: Ah. I think I see what you're saying now: it's about the resulting data size, not just out-of-bounds indexes. Yes, you can end up with significantly more than 1 MB of data, especially if you have CJK text or the like in your data.

Yeah, truncating the BINARY before transcoding would let you be exact. (Or, forcing the UTF-8 string back to BINARY, truncating, and then re-transcoding, to keep the control flow the same.) Probably simple. Do we care enough to do this now?

Anyone know where the 1 MB limit comes from, anyway?

Comment (apjanke) [inline Library/Homebrew/cmd/test-bot.rb]: (That's why I changed the message from `1MB` to `~1MB`, to indicate the size was only approximate.)

Comment (apjanke) [inline Library/Homebrew/cmd/test-bot.rb]: Know any formulae whose output is long enough to trigger the truncation, but still compile pretty quickly?

Comment (UniqMartin) [inline Library/Homebrew/cmd/test-bot.rb]: I don't have answers for most of the questions. Just wanted to make you ware of a potential issue.

> Do we care enough to do this now?

Probably not …

Comment (apjanke) [inline Library/Homebrew/cmd/test-bot.rb]: Thanks. I was aware of it (mentioned it in the original issue post), just didn't think it was worth taking care of. But the "truncate in BINARY and transcode" fix you suggested is pretty straightforward, so I'm going to go ahead and try it here.

Comment (UniqMartin) [inline Library/Homebrew/cmd/test-bot.rb]: > I was aware of it (mentioned it in the original issue post), just didn't think it was worth taking care of.

Sorry! I think I missed this somehow.

Comment (apjanke) [inline Library/Homebrew/cmd/test-bot.rb]: Cool. Anyway, I altered it to use your approach. Got long enough I split it out to a separate `sanitize_for_xml` method lower down in the file. Look about right to you? I'm going to go looking for a long-output formula to test it on.

File: Library/Homebrew/cmd/test-bot.rb [hunk 2]
```diff
@@ -981,4 +983,35 @@ def test_bot
...
+        output = output.gsub(invalid_xml_pat, "\uFFFD")
+      else
+        output = output.delete("\000\a\b\e\f\x2\x1f")
+      end
+
+      # Truncate to 1MB
+      if output.bytesize > BYTES_IN_1_MEGABYTE
+        if ruby_has_encoding?
+          binary_output = output.force_encoding("BINARY")
+          binary_output = binary_output.slice(-BYTES_IN_1_MEGABYTE, BYTES_IN_1_MEGABYTE)
```
Comment (UniqMartin) [inline Library/Homebrew/cmd/test-bot.rb]: This should be assigning back to `output` (not `binary_output`) I guess …

File: Library/Homebrew/cmd/test-bot.rb [hunk 3]
```diff
@@ -981,4 +983,35 @@ def test_bot
 
     Homebrew.failed = any_errors
   end
+
+  def sanitize_output_for_xml(output)
+    unless output.empty?
```
Comment (UniqMartin) [inline Library/Homebrew/cmd/test-bot.rb]: You could make that `return "" if output.empty?` and reduce indentation in the remainder of the method.

File: Library/Homebrew/cmd/test-bot.rb (hunk 2 above)
Comment (apjanke) [inline Library/Homebrew/cmd/test-bot.rb]: Fixed; thanks.

File: Library/Homebrew/cmd/test-bot.rb (hunk 3 above)
Comment (apjanke) [inline Library/Homebrew/cmd/test-bot.rb]: I don't know if that's Homebrew's house style? I personally like to take early returns to keep indentation/nesting low too, but most of the `brew` code base doesn't seem to do that. Is there some guidance on that?

File: Library/Homebrew/cmd/test-bot.rb [hunk 4]
```diff
@@ -981,4 +983,35 @@ def test_bot
...
+      # Remove invalid XML CData characters from step output.
+      if ruby_has_encoding?
+        # This is the regex for valid XML chars, but only works in Ruby 2.0+
+        # /[\x09\x0A\x0D\x20-\uD7FF\uE000-\uFFFD\u{10000}-\u{10FFFF}]/
+        # For 1.9 compatibility, use the inverse of that, which stays under \u10000
```
Comment (mistydemeo) [inline Library/Homebrew/cmd/test-bot.rb]: This is the nittiest nitpick, but 1.9 was the version that introduced encodings - 1.8 is the one without.

File: Library/Homebrew/cmd/test-bot.rb [hunk 5]
```diff
@@ -981,4 +983,35 @@ def test_bot
...
+        # invalid_xml_pat = /[\x00-\x08\x0B\x0C\x0E-\x1F\uD800-\uDFFF\uFFFE\uFFFF]/
+        # But Ruby won't allow you to reference surrogates, so we have:
+        invalid_xml_pat = /[\x00-\x08\x0B\x0C\x0E-\x1F\uFFFE\uFFFF]/
```
Comment (mistydemeo) [inline Library/Homebrew/cmd/test-bot.rb]: `\u` escapes don't exist in 1.8, so `\uFFFE` is coming out as the literal text `uFFFE`.

Comment (mistydemeo) [inline Library/Homebrew/cmd/test-bot.rb]: Arg, ignore this, I completely ignored the conditional here.

File: Library/Homebrew/cmd/test-bot.rb (hunk 4 above)
Comment (apjanke) [inline Library/Homebrew/cmd/test-bot.rb]: Yes, sorry if my comment here isn't clear: I mean that while 1.9 introduced encodings and `\u` escapes, only 2.0 introduced the `\u{...}` escapes that allow more than 4 digits for representing values U+10000 and above. So the original pattern `/[\x09\x0A\x0D\x20-\uD7FF\uE000-\uFFFD\u{10000}-\u{10FFFF}]/` blows up in 1.9. I inverted the pattern so it could be written with just the `\uXXXX` escapes and not `\u{...}`, so it would work in 1.9, too.

Maybe it could be worded better?

**Comment by UniqMartin** [comment]
@apjanke There was still an issue with `sanitize_output_for_xml` (attempting to access the private `fix_encoding!` method of the `Step` class; could be observed in [this job](http://bot.brew.sh/job/Homebrew%20Versions%20Pull%20Requests/1315/version=el_capitan/console)) that I tried to (somewhat carelessly) fix in a2c23dfec569c6e73d90cb20c7d4c26cced258d5, which killed a whole bunch of jobs and needed more fixing in 37b817ed394986808d2fdb595c272ede7a67f20d. Would be nice if you could double-check my “fix” and see if that needs any cleaning up.

At least now the truncation also works as it should as can be seen in [this second run of the above mentioned job](http://bot.brew.sh/job/Homebrew%20Versions%20Pull%20Requests/1317/version=el_capitan/console).

----

Relevant parts of the logs of the above mentioned jobs (as the links won't stay valid forever):

```
[…]
==> brew cleanup --prune=7
==> git clean -ffdx
HEAD is now at a9c0361 audit: require plist_options when using plist
Removing Cellar/
Removing Library/Homebrew/test/.bundle/
Removing Library/Homebrew/test/fs_leak_log
Removing Library/Homebrew/test/vendor/
Removing etc/openssl/
Removing share/info/
Removing share/man/whatis
Removing var/
Error: undefined method `fix_encoding!' for Homebrew:Module
Please report this bug:
    https://git.io/brew-troubleshooting
/usr/local/Library/Homebrew/cmd/test-bot.rb:1007:in `sanitize_output_for_xml'
/usr/local/Library/Homebrew/cmd/test-bot.rb:955:in `block (2 levels) in test_bot'
/usr/local/Library/Homebrew/cmd/test-bot.rb:948:in `each'
/usr/local/Library/Homebrew/cmd/test-bot.rb:948:in `block in test_bot'
/usr/local/Library/Homebrew/cmd/test-bot.rb:943:in `each'
/usr/local/Library/Homebrew/cmd/test-bot.rb:943:in `test_bot'
/usr/local/Library/brew.rb:84:in `<main>'
Build step 'Execute shell' marked build as failure
Recording test results
ERROR: Step ‘Publish JUnit test result report’ failed: No test report files were found. Configuration error?
Archiving artifacts
Started calculate disk usage of build
Finished Calculation of disk usage of build in 0 seconds
Started calculate disk usage of workspace
Finished Calculation of disk usage of workspace in 0 seconds
[WS-CLEANUP] Deleting project workspace...[WS-CLEANUP] done
Finished: FAILURE
```

```
[…]
==> brew cleanup --prune=7
==> git clean -ffdx
HEAD is now at 37b817e test-bot: fix access to relocated method
Removing Cellar/
Removing Library/Homebrew/test/.bundle/
Removing Library/Homebrew/test/fs_leak_log
Removing Library/Homebrew/test/vendor/
Removing etc/openssl/
Removing share/info/
Removing var/
Build step 'Execute shell' marked build as failure
Recording test results
Archiving artifacts
Started calculate disk usage of build
Finished Calculation of disk usage of build in 0 seconds
Started calculate disk usage of workspace
Finished Calculation of disk usage of workspace in 0 seconds
[WS-CLEANUP] Deleting project workspace...[WS-CLEANUP] done
Finished: FAILURE
```

File: Library/Homebrew/cmd/test-bot.rb (hunk 1 above)
Comment (mikemcquaid) [inline Library/Homebrew/cmd/test-bot.rb]: > Anyone know where the 1 MB limit comes from, anyway?

Either Jenkins and/or Travis CI. I'd be happy with changing it to anything that's "definitely under 1MB" even if that assumes all characters are double-width unicode, for instance.

**Comment by apjanke** [comment]
New fix looks good. Thanks! Don't know how I missed that when pushing; I ran test-bot locally before doing the merge and it seemed fine.

**Comment by apjanke** [comment]
WRT the 1 MB limit: the truncation calculation is now I believe correct, so it will truncate to exactly 1 MB of output. But it will also slap the phrase "output truncated to 1MB:" on it, so it'll end up slightly over 1 MB. If we want a margin of error we could dial it down to 900K or so.

**Comment by mikemcquaid** [comment]
> WRT the 1 MB limit: the truncation calculation is now I believe correct, so it will truncate to exactly 1 MB of output. But it will also slap the phrase "output truncated to 1MB:" on it, so it'll end up slightly over 1 MB. If we want a margin of error we could dial it down to 900K or so.

Haha. Yeh, that sounds good 👍

**Comment by apjanke** [comment]
Done in #39.
